src/coprocessor/statistics/fmsketch.rs

```diff
@@ -0,0 +1,187 @@
+        self.mask = mask;
+        let to_remove: Vec<u64> = self.hash_set
+            .iter()
+            .filter(|x| (*x & self.mask) != 0)
+            .cloned()
+            .collect();
+        for x in &to_remove {
+            self.hash_set.remove(x);
+        }
+    }
+
+    fn insert_hash_value(&mut self, hash_val: u64) {
+        if (hash_val & self.mask) != 0 {
+            return;
+        }
+        self.hash_set.insert(hash_val);
+        if self.hash_set.len() > self.max_size {
+            let mask = (self.mask << 1) | 1;
+            self.update_mask(mask);
+        }
+    }
+
+    fn insert(&mut self, v: &Datum) -> Result<()> {
+        let bytes = try!(datum::encode_value(as_slice(v)));
+        let hash = {
+            let mut hasher = FnvHasher::default();
+            hasher.write(&bytes);
+            hasher.finish()
+        };
+        self.insert_hash_value(hash);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    struct TestData {
+        samples: Vec<Datum>,
+        rc: Vec<Datum>,
+        pk: Vec<Datum>,
+    }
+
+    fn generate_samples(count: usize) -> Vec<Datum> {
+        let mut samples = Vec::with_capacity(count);
+        samples.resize(count, 0);
+        let start = 1000;
+
+        for mut item in samples.iter_mut().take(start).skip(1) {
+            *item = 2;
+        }
+
+        for (id, mut item) in samples.iter_mut().enumerate().take(count).skip(start) {
+            *item = id;
+        }
+
+        let mut id = start;
+        while id < count {
+            samples[id] += 1;
+            id += 3;
+        }
+
+        id = start;
+        while id < count {
+            samples[id] += 2;
+            id += 5;
+        }
+        samples.into_iter().map(|v| Datum::I64(v as i64)).collect()
+    }
+
+    impl Default for TestData {
+        fn default() -> TestData {
+            let samples = generate_samples(10000);
+            let count = 100000;
+            let rc = generate_samples(count);
+
+            let mut pk = Vec::with_capacity(count);
+            pk.resize(count, Datum::Null);
+            for (id, mut item) in pk.iter_mut().enumerate().take(count) {
+                *item = Datum::I64(id as i64);
+            }
+            TestData {
+                samples: samples,
+                rc: rc,
+                pk: pk,
+            }
+        }
+    }
+
+    #[test]
+    fn test_sketch() {
+        // This test was ported from tidb.
+        let max_size = 1000usize;
+        let data = TestData::default();
+        let mut sample = FMSketch::build(&data.samples, max_size).unwrap();
+        assert_eq!(sample.ndv(), 6624);
+        let rc = FMSketch::build(&data.rc, max_size).unwrap();
+        assert_eq!(rc.ndv(), 74240);
+        let pk = FMSketch::build(&data.pk, max_size).unwrap();
+        assert_eq!(pk.ndv(), 99968);
+        sample.merge(&pk);
+        sample.merge(&rc);
+        assert_eq!(sample.ndv(), 99968);
```

Comment:
It looks strange to me using `assert_eq` for an approximate result.

Reply:
I use `assert_eq` here in order to make sure its implement is same with `tidb`.